### ISSUE TITLE
[DISCUSSION] GEOS-Chem pressures and water vapor

### ISSUE THREAD

**Issue by yantosca** [geoschem/geos-chem#458]
@cdholmes wrote:

>There appear to be some subtle problems in the way that the dry air pressures, masses, and box sizes are calculated in calc_met_mod.F90. Difference from the current approach are likely to be small, but could matter in very humid regions. I noticed these things when trying to resolve an issue that @chfite  was having. 
>
>I can fix all of these issues, but I want to make sure there aren’t other unintended consequences. What transport and mass conservation tests can we run to check the results? 
>
>1. The dry air mass is defined in calc_met_mod.F90 in a way that is not consistent with the total air mass (or pressure) and specific humidity. The calculation can also be much simpler, as shown on these few lines.

```
! Total air mass
ADwet = Area * ( Pbottom – Ptop ) / g0
! Dry air mass
AD = ADwet * ( 1 – SPHU * AIRMW / H2OMW )
```

>  Most of the remaining issues are just consequences of item 1. 
>
>2. The partial pressures of dry air and water vapor should be defined from total pressure and specific humidity (not using RH) to guarantee consistency with the mass distribution in the model. [The current approach will give the same result if and only if the Esat defined in GC is the same as GEOS-5. Switching to SPHU will ensure that GC remains consistent regardless of what GEOS-5 does now or if it changes in the future.]

```
! Correct approach
Ev = P * ( SPHU * AIRMW / H2OMW )
PDRY = P - Ev
```

>3. The virtual temperature (TV) used to determine grid box height and volume should be based on conditions at the grid cell center, not at the edge as currently done.  
>
>4. The pressure edges at edge L is currently defined by extrapolating DOWN from the midpoint of level L. You would get a different edge pressure if you extrapolated UP from the midpoint of level L-1. That doesn’t seem right. Unless there is a reason for this that I can’t see, this seems like a bug. I think the correct method is to use the dry air mass as specified above to calculate dry air pressure. 
>
>5. Potential temperature is typically defined at a fixed reference temperature (often 1000 hPa) not at surface pressure. GC currently uses surface pressure. For some applications, that might be OK, but as far as I can see, the potential temperature is only used in planeflight_mod, so it should definitely be set at a fixed temperature.
>
>6. The things above led me to notice that GET_DELP_DRY in pressure_mod is using the A’s and B’s to calculate the dry pressure thickness of a grid cell (effectively the dry air mass). That can’t be correct. The A’s and B’s either apply to dry pressure or total (wet) pressure, but not both (I don’t know which they were applied to in GEOS-5 but I assume wet).

**Comment by yantosca** [comment]
We are currently prioritizing the 13.0.0 release this fall.  More realistically any updates regarding water vapor and pressure would go into an early 13 series release.

**Comment by yantosca** [comment]
Also note: changing treatment of water pressure may have implications for connecting GEOS-Chem to external models such as CESM, GEOS, etc. and should be studied with more care before implementation.   For the 13.0.0 release we are trying not to add sweeping science changes into the code given the structural reordering of the code base that has been necessary.

**Comment by lizziel** [comment]
Hi @cdholmes, as Bob said we need to prioritize getting 13.0.0 ready for benchmarking before tackling this. In the meantime, you can read documentation of the current implementation [here](http://wiki.seas.harvard.edu/geos-chem/index.php/Moisture_Updates_in_v11-01).

**Comment by cdholmes** [comment]
Thanks Lizzie and Bob. How did you test the water vapor imprinting on tracers previously? Do you have some recommended transport and mass conservation tests that I should run before and after making changes?

Christopher D. Holmes, PhD
Earth, Ocean and Atmospheric Science
Florida State University
http://acgc.eoas.fsu.edu
________________________________

**Comment by lizziel** [comment]
Hi @cdholmes, unfortunately our previous tool for mass conservation testing was retired as part of binary diagnostics removal earlier in version 12. However, there are a few things you can still do depending on what level of validation you are doing.

For a  general test you can use the transport tracer simulation, output periodic restart files via the restart collection in HISTORY.rc, and calculate global mass for the passive tracer using the restart file values. All meteorology needed for the conversion to mass from mol/mol is contained with the file. You can compare start and end global mass this way, or look for a trend from a time series of restart output.

If you want to test mass conservation run-to-run as you develop, and using very short runs (which btw is what I did when I put in the moisture updates in v11.02), then I recommend putting in prints in main.F. I added the subroutine Print_Global_Species_Kg exactly for this purpose. See https://github.com/geoschem/geos-chem/blob/b95926b8b53d27817090c36e1c9f939110ce03f1/GeosUtil/unitconv_mod.F90#L352-L468.

Admittedly that subroutine needs some polishing, but you can use it to print global mass, along with mass and met variables for a single grid box (useful when looking at moisture updates), with some customization. This block is specifically what you can customize based on what you want to see: https://github.com/geoschem/geos-chem/blob/b95926b8b53d27817090c36e1c9f939110ce03f1/GeosUtil/unitconv_mod.F90#L442-L453. The subroutine takes a character string to keep track of which call it is printing values for, e.g. "before do_chemistry".

My approach was to put a call to this subroutine at the start of each dynamic loop, run for a few timesteps, and see if anything looked amiss in terms of mass change. It is also useful to put this before and after calls to AirQnt (or components that call it) since that subroutine updates the mixing ratio to preserve mass across changes in moisture. I have found that this is the most direct way to debug mass conservation problems introduced by changes in moisture equations. You can also print global mass between each component call in main.F, but beware that some fluctuation is expected due to pressure corrections applied in advection.

Let me know if you have any questions! We plan on adding a mass conservation table to GCPy soon for use during transport tracer benchmarks, but also as a general mass test tool.

**Comment by cdholmes** [comment]
Pull request #494 fixes the issues with saturation vapor pressure and potential temperature that I mentioned earlier. 

There are still some suspicious features of the way that dry air pressure is computed for use in transport (tpcore).
The dry surface pressure is (correctly) calculated by removing all water vapor from the column. 
However, the dry pressure edges and thickness (GET_PEDGE_DRY, GET_DELP_DRY in pressure_mod) are then computed from the dry surface pressure using the Hybrid A's and B's. With this approach, the dry pressure vertical profile depends only on the total amount of water vapor in the column, ignoring where it is distributed. 
In other words, the dry pressure profile currently used for advection is /inconsistent/ with the specific humidity profile.
I'm not sure how to fix this because TPCORE is designed around the assumption that the pressure profile is based on the Hybrid A's and B's, but dry pressure does not follow this distribution.
It's nice that GEOS-Chem transport currently conserves mass, but I'm not convinced that it is doing so for the right reasons.

**Comment by lizziel** [comment]
See also PR https://github.com/geoschem/geos-chem/pull/495 from @cdholmes which supercedes https://github.com/geoschem/geos-chem/pull/494.
